kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/bind.kt

```diff
@@ -1,48 +1,116 @@
 package io.kotest.property.arbitrary
 
-fun <A, T> Arb.Companion.bind(arbA: Arb<A>, createFn: (A) -> T): Arb<T> = arb {
-   createFn(arbA.single(it))
-}
+import io.kotest.property.RandomSource
+import io.kotest.property.Sample
+
+fun <A, T> Arb.Companion.bind(arbA: Arb<A>, createFn: (A) -> T): Arb<T> = arbA.map(createFn)
+
+fun <A, B, T> Arb.Companion.bind(arbA: Arb<A>, arbB: Arb<B>, createFn: (A, B) -> T): Arb<T> = object : Arb<T> {
+   override fun edgecases(): List<T> = emptyList()
+
+   override fun samples(rs: RandomSource): Sequence<Sample<T>> = values(arbA.samples(rs), arbB.samples(rs))
 
-fun <A, B, T> Arb.Companion.bind(arbA: Arb<A>, arbB: Arb<B>, createFn: (A, B) -> T): Arb<T> = arb {
-   val a = arbA.single(it)
-   val b = arbB.single(it)
-   createFn(a, b)
+   override fun generate(rs: RandomSource): Sequence<Sample<T>> = values(arbA.generate(rs), arbB.generate(rs))
+
+   private fun values(aSamples: Sequence<Sample<A>>, bSamples: Sequence<Sample<B>>): Sequence<Sample<T>> {
+      val a = aSamples.iterator()
+      val b = bSamples.iterator()
+
+      return sequence {
+         while (a.hasNext() && b.hasNext()) {
+            yield(Sample(createFn(a.next().value, b.next().value)))
+         }
+      }
+   }
 }
 
 fun <A, B, C, T> Arb.Companion.bind(
    arbA: Arb<A>,
    arbB: Arb<B>,
    arbC: Arb<C>,
    createFn: (A, B, C) -> T
-): Arb<T> = arb {
-   val a = arbA.single(it)
-   val b = arbB.single(it)
-   val c = arbC.single(it)
-   createFn(a, b, c)
+): Arb<T> = object : Arb<T> {
+   override fun edgecases(): List<T> = emptyList()
+
+   override fun samples(rs: RandomSource): Sequence<Sample<T>> = values(arbA.samples(rs), arbB.samples(rs), arbC.samples(rs))
+
+   override fun generate(rs: RandomSource): Sequence<Sample<T>> = values(arbA.generate(rs), arbB.generate(rs), arbC.generate(rs))
+
+   private fun values(
+      aSamples: Sequence<Sample<A>>,
+      bSamples: Sequence<Sample<B>>,
+      cSamples: Sequence<Sample<C>>
+   ): Sequence<Sample<T>> {
+      val a = aSamples.iterator()
+      val b = bSamples.iterator()
+      val c = cSamples.iterator()
+
+      return sequence {
+         while (a.hasNext() && b.hasNext() && c.hasNext()) {
+            yield(Sample(createFn(a.next().value, b.next().value, c.next().value)))
+         }
+      }
+   }
 }
```

Comment:
What's the advantage here, because it's going to propogate the random into the other arbs anyway? Is it just to avoid returning a new sequence each time that may have some performance penalty ?

Reply:
There may be a performance advantage. However, the main reason I have done this is that the prior implementation results in issues for certain types of arbs. If the arb does any sort of initialization of the sequence, then the initialization occurs for each result of the bind.

For example, the following test would output duplicate values because `distinct` sets up the set it collects results in each call.
```kotlin
forAll(100, Arb.bind(Arb.int(0, 10).distinct()) { it }) {
  println(it)
  true
}
```

The `regex` arb had the same issue before 711898e133bb5088c3cbe7cf250d9e8eb2ec91e2. The regex would always emit the same value to the bind because the Generex was reset each call to samples.

Another issue that I found is that `bind` was only calling `samples`, so it would never receive the edgecases from the arb. Maybe this was intended? If this is the intended behavior, it seems unexpected to me, but I can change my PR to call samples instead of generate.

Reply:
All seems good thanks :)